README.md
# Cloud Haoma

The project was done in a team effort. Team Members:-
1. Supriya Aggarwal 2014A7PS0013P
2. Vidhi Jain 2014A7TS0113P
3. Abhishek Gupta 2014A7PS0026P

Run the following commands on all the nodes of the cluster

Python version 2.7

1. sudo apt-get install nodejs nodejs-legacy npm libzmq-dev git
2. pip install pyzmq-static
2. Open Terminal in the base directory of the project
3. npm install
4. pip install zmq


Network topology is star.

Run the following on the designated router.

1. python2.7 router.py 12345 [[1,2,3],[4,5]] '[6]' false


Run the following on the cluster nodes (you may include router also (Let IPADDRESS of router node for this be 127.0.0.1):-

Group I 
1. python2.7 server.py 1 tcp://127.0.0.1 12345 ["1","2","3"] on node1
2. python2.7 server.py 2 tcp://127.0.0.1 12345 ["1","2","3"] on node2
3. python2.7 server.py 3 tcp://127.0.0.1 12345 ["1","2","3"] on node3

Group II
1. python2.7 server.py 4 tcp://127.0.0.1 12345 ["4","5"] on node4
2. python2.7 server.py 5 tcp://127.0.0.1 12345 ["4","5"] on node5



Run the following code on the node which you want to run your web server
(Currently webserver tries to insert hardcoded values in the db on receiving
 the API call instead of using request Data)


1.node clientServer.js 6 tcp://127.0.0.1 12345 true 3000

API CALL:- 

IPADDRESSOFWEBSERVER:3000/readData


//For writing request
//Requests Information:-
//POST REQUEST - multipart/formdata
//Keys - fileData, clientId, sourceId

IPADDRESSOFWEBSERVER:3000/pushData


//For reading request
//Request Infromation:
//Post Request - multipart/formdata
//Keys - clientId, sourceId

IPADDRESSOFWEBSERVER:3000/readData




# Strategy
Using RAFT for log replication. 
Each log entry contains the write request by a client. Request contains sufficient 
information to identify the file. State machine operation is to append the request 
data to the identified file.
Reads can be handled by the leader
Data is sharded across the groups (here gropu I and II). Raft replicates the logs on all the nodes of the cluster => So if we run a single Raft instance for our cluster, we can't achieve a replication factor less than that of cluster size => Cluster should be divided into groups(shards) and raft instance should run for that group. => A data item will be present in one group only and replication is within the nodes of that group

Some consensus algorithms support partial replication in which we can control which data will replicated and its replication factor. Raft doesn't support it so we should shard our data to get decrease replication factor. Also, sharding improves writes throughput.

# Status
Code for supporting the write and read operation is working with the API Calls.

# Reference

This project is based on the raft implementation present at https://github.com/albertdb/Raft.

The entire code was rewritten into python (Original code was written in nodejs which
we suspect to be better choice than python because of its asynchronous nature)

ZeroMQ (http://zeromq.org/) is used for providing communication



changes.md
1. BugFix - Made recoveryPrevIndex global in appendEntries function
2. Reading filename and data from request object.
	2.1 Request changed from get to post
	2.2. Only requests with content-type = "multipart/form-data" is supported


log.py

class LogEntry(object):
	def __init__(self, clientId, requestId, data, term):
		self._clientId = clientId
		self._requestId = requestId
		self._data = data
		self._term = term

	def __repr__(self):
		return "\nclientId: "+ str(self._clientId) + \
		"\nrequestId: " + str(self._requestId) + \
		"\ndata: " + str(self._data) + \
		"\nterm: " + str(self._term) + "\n "

class Log(object):
	def __init__(self, server_id):
		self._firstIndex = 0
		self._length = 1;
		self._entries = [LogEntry(-1, 0, None, 0), ]

	def __repr__(self):
		return "\nfirstIndex " + str(self._firstIndex) + \
				"\nlength " + str(self._length) + \
				"\nEntries" + str(self._entries)

	def push(self, value):
		self._length += 1
		print value, type(value)
		trans = LogEntry(value['_clientId'], value['_requestId'], value['_data'], value['_term'])
		self._entries.append(trans)

	def pop(self):
		self._length -= 1
		self._entries.pop()

	def shift(self):
		pass

	def slice(self, from1, to):
		return self._entries[from1:to]


router.py
import sys
import zmq
import json
import ast

#python2.7 router.py 12345 [[1,2,3],[4,5]] '[6]' '[5002, 5003, 5004]' false
#            0         1           2         3          4                5	


#Sharding Function
def getGroup(customerID, numGroups):
	return hash(customerID) % numGroups;

context = zmq.Context()
#All clusterNodes will send their packets to this socket
receiver_socket = context.socket(zmq.PULL)
receiver_socket.bind("tcp://*:"+str(sys.argv[1]))

#All clusterNodes will receive packets from this socket
sender_socket = context.socket(zmq.PUB)
sender_socket.bind("tcp://*:5000");


#NodeIds of all clusterServers
groups = ast.literal_eval(sys.argv[2]) #list of lists
numGroups = len(groups)
servingIndices = [0] * numGroups

#NodeIds of all webServers (currently only one webServer supported)
webServers = ast.literal_eval(sys.argv[3])
webServers = [str(x) for x in webServers]
clusterServerId = 0

#WebServer receives response from this socket
#Port numbers of all webServers on which they will listen
webServersPortNos = ast.literal_eval(sys.argv[4])
print webServersPortNos	
webServersPortNos = [str(x) for x in webServersPortNos]
port_dict = {}
if len(webServersPortNos) != len(webServers):
	print "Number of port numbers doesn't match number of webservers!\n"
	exit()

for i in range(len(webServers)):
	socket = context.socket(zmq.PUSH)
	socket.bind("tcp://*:"+webServersPortNos[i])
	port_dict[webServers[i]] = socket



#Whether to print
debug = (sys.argv[5] == "true")

if debug:
	print "Groups = ", groups
	print "webServers = ", webServers

while True:
	data = receiver_socket.recv_json()
	dest_id = str(data.get('dest'))
	#if debug:
	#	print data
	#print dest_id
	if dest_id == 'None': 
		#Packet received from a webserver
		#We don't know the destination
		#Find group from customerId
		#Assign in a round robin fashion in the group
		customerId = data['customerId']
		groupId = getGroup(customerId, numGroups)
		dest_id = str(groups[groupId][servingIndices[groupId]])
		print "Serving to group %d and destination %s"%(groupId, dest_id)
		servingIndices[groupId] = (servingIndices[groupId]+1)%(len(groups[groupId]))
	if dest_id not in webServers:
		sender_socket.send("%s %s"%(dest_id, data))
	else:
		print data
		port_dict[dest_id].send("%s"%(json.dumps(data)))


server.py
import zmq
import sys
import ast
import threading
import json
import pickle
from log import LogEntry,Log
from utility import *
import random

# global variables
currentTerm = 0
state = 'f'
votedFor = None
lastKnownLeaderID = None
commitIndex = 0
maybeNeedToCommit = False
lastApplied = 0
nextIndex = {}
matchIndex = {}
recoveryMode = False
recoveryPrevLogIndex = 0
grantedVotes = 0
election = random.randint(150, 300)
heartbeatTime = 100
persistentStoreTime = 2000
commitTime = 500
shift = False
shiftHeart = False
electionTimeCall = False
heartbeatTimeCall = False
#python2.7 server.py 1 tcp://127.0.0.1 12345 ["1","2","3"]
context = zmq.Context()

server_id = int(sys.argv[1])
router_address = sys.argv[2]
port_no = sys.argv[3]
port_no2 = '5000'
clusterMembers = ast.literal_eval(sys.argv[4])
clusterMembers = [int(x) for x in clusterMembers]
sender_socket = context.socket(zmq.PUSH)
sender_socket.connect(router_address +":"+ port_no)

receiver_socket = context.socket(zmq.SUB)
receiver_socket.setsockopt(zmq.SUBSCRIBE, str(server_id))
receiver_socket.connect(router_address + ":" +port_no2)

log = Log(server_id)

for member in clusterMembers:
	matchIndex[member] = 0
	nextIndex[member] = 1

def sendMessage(destid, msg):
	msg['dest'] = destid
	sender_socket.send_json(msg)

def electionTimeout():
	global currentTerm, electionTimeCall, state, votedFor, grantedVotes, clusterMember, lastKnownLeaderID, log
	global server_id, shift, election
	if electionTimeCall == True:
		if lastKnownLeaderID == None:
			print("No known leader for me, starting election!")
		else:
			print("No heartbeat received from leader, starting new election!")
		currentTerm += 1
		state = 'c'
		votedFor = server_id
		grantedVotes = 1
		msg = {'rpc':'requestVote'
		, 'term':currentTerm
		, 'candidateId':server_id
		, 'lastLogIndex':log._length-1
		, 'lastLogTerm':log._entries[-1]._term};
		for destid in clusterMembers:
			if destid != server_id:
				sendMessage(destid, msg)
	if shift:
		electionTimeCall = True
		shift = False
	threading.Timer(election/1000.0, electionTimeout).start()


def heartbeatTimeout():
	global heartbeatTimeCall, shift, currentTerm, server_id, shiftHeart, heartbeatTime
	global electionTimeCall, state, votedFor, grantedVotes, clusterMember, lastKnownLeaderID, log
	if heartbeatTimeCall == True:
		msg = {'rpc':'appendEntries'
		, 'term':currentTerm
		, 'leaderId': server_id
		, 'prevLogIndex':log._length-1
		, 'prevLogTerm':log._entries[-1]._term
		, 'entries' : []
		, 'leaderCommit':commitIndex};
		for destid in clusterMembers:
			if destid != server_id:
				sendMessage(destid, msg)
		electionTimeCall = False
		shift = False
	if shiftHeart == True:
		heartbeatTimeCall = True
		shiftHeart = False
	threading.Timer(heartbeatTime/1000.0, heartbeatTimeout).start()

def persistentStoreTimeout():
	global currentTerm, votedFor, log, lastApplied
	# print("here\n")
	writeToPersistentStore(currentTerm, votedFor, log, lastApplied)
	threading.Timer(persistentStoreTime/1000.0, persistentStoreTimeout).start()

def newNullEntry():
	global server_id, currentTerm, clusterMembers, server_id, log
	dummy = LogEntry(-1, 0, None, currentTerm)
	msg = {'rpc':'appendEntries'
		, 'term':currentTerm
		, 'leaderId':server_id
		, 'prevLogIndex':log._length-1
		, 'prevLogTerm':log._entries[log._length-1]._term
		, 'entries': [dummy.__dict__]
		, 'leaderCommit':commitIndex}
	print clusterMembers, server_id
	for clusterMember in clusterMembers:
		if clusterMember != server_id:
			if nextIndex[clusterMember] == log._length:
				nextIndex[clusterMember] += 1
				sendMessage(clusterMember, msg)
	log.push(dummy.__dict__)


#Bad!! TODO: make this async IO
def processEntries(upTo):
	global server_id, state,lastApplied, log
	for entryIndex in range(lastApplied, upTo):
		entry = log._entries[entryIndex]
		if state == 'l' and entry._clientId!=-1:
			clientId = entry._clientId
			requestId = entry._requestId
			msg = {'dest':clientId
			, 'requestId':requestId
			, 'status' : 'Success'
			, 'rpc' : 'addEntryReply'};
			print "After processing : %s"%(msg)
			sendMessage(clientId, msg);
		# print entry.__dict__, type(entry)
		# print entry._data, type(entry._data)
		write_to_file(entry._data)
	lastApplied = upTo



def requestVote(term, candidateId, lastLogIndex, lastLogTerm):
	global heartbeatTimeCall, shift, currentTerm, electionTimeCall, state, votedFor, grantedVotes, clusterMember, lastKnownLeaderID, log
	msg = []
	if term >= currentTerm:
		if term > currentTerm : #I am in the past
			print("Election in progress!")
			currentTerm = term
			if (state == 'l'):
				print("Demoting to follower state.")
			state = 'f'
			votedFor = None
			recoveryMode = None
			heartbeatTimeCall = False
		if (votedFor == None) \
		and (log._length==log._firstIndex or lastLogTerm >  log._entries[-1]._term or (lastLogTerm == log._entries[-1]._term and lastLogIndex>=log._length-1)):
			votedFor = candidateId
			print("Vote given to candidate %s"%(candidateId))
			msg = {'rpc':'replyVote', 'term':currentTerm, 'voteGranted':True}
			electionTimeCall = False #Reset election Timeout
			shift = True
		else:
			msg = {'rpc':'replyVote', 'term':currentTerm, 'voteGranted':False};
	else:  #Sender is In the Past
		msg = {'rpc':'replyVote', 'term':currentTerm, 'voteGranted':False};
	sendMessage(candidateId, msg)


def replyVote(term, voteGranted):
	global heartbeatTimeCall, currentTerm, electionTimeCall, state, votedFor
	global server_id, grantedVotes, clusterMember, lastKnownLeaderID, log
	if term > currentTerm:
		currentTerm = term
		state = 'f'
		grantedVotes = 0
		votedFor = None
	elif voteGranted==True and term==currentTerm and state=='c':
		grantedVotes += 1
		print("Received vote")
		if grantedVotes > len(clusterMembers)/2:
			print("Election win. Say Hi to new leader.")
			state = 'l'
			lastKnownLeaderID = server_id
			grantedVotes = 0
			# print "default nextIndex = %d and matchIndex = %d"%(log._length,log._length-1)
			for destid in clusterMembers:
				if destid != server_id:
					nextIndex[destid] = log._length
					matchIndex[destid] = log._length-1
			heartbeatTimeCall = True
			electionTimeCall = False
			newNullEntry()


def appendEntries(term, leaderId, prevLogIndex, prevLogTerm, entries, leaderCommit):
	global currentTerm, electionTimeCall, lastKnownLeaderID, currentTerm, log, shift
	global recoveryMode, commitIndex, server_id, recoveryPrevLogIndex
	processEn= False
	msg = {}
	if term>=currentTerm:
		electionTimeCall = False
		shift = False
		if (lastKnownLeaderID != leaderId):
			print("Election result %d is the new leader"%(leaderId))
		lastKnownLeaderID = leaderId
		if term > currentTerm :
			currentTerm = term
		log_length = log._length
		log_firstIndex = log._firstIndex
		if prevLogIndex < log_length:
			log_prevTerm = log._entries[prevLogIndex]._term

		if (prevLogIndex < log_length and (log_length==log_firstIndex or log_prevTerm==prevLogTerm)):
			if recoveryMode:
				print("Last matching entry found. Exiting recovery mode.!")
			recoveryMode = False
			for entry in entries:
				log.push(entry)
			msg = {'rpc':'replyAppendEntries'
			, 'term':currentTerm
			, 'followerId':server_id
			, 'entriesToAppend': len(entries)
			, 'success':True};
			if leaderCommit > commitIndex:
				commitIndex = min(leaderCommit, log._length)
				processEn = True
			electionTimeCall = False #Reset electionTimeOut
			shift = True
		elif recoveryMode==False or (recoveryMode and prevLogIndex < recoveryPrevLogIndex):
			while log._length > prevLogIndex:
				log.pop()
			if recoveryMode == False:
				print("Log is outdated. Entering recovering mode.")
			recoveryMode = True
			recoveryPrevLogIndex = prevLogIndex
			msg = {'rpc':'replyAppendEntries'
			, 'term':currentTerm
			, 'followerId':server_id
			, 'entriesToAppend': prevLogIndex
			, 'success':False};
	else:
		msg = {'rpc':'replyAppendEntries'
		, 'term':currentTerm
		, 'followerId':server_id
		, 'entriesToAppend': prevLogIndex
		, 'success':False};
	if msg != {}:
		# print ("leaderid = %s"%(leaderId))
		sendMessage(leaderId, msg)
	if processEn:
		processEntries(commitIndex)


def replyAppendEntries(term, followerId, entriesToAppend, success):
	global currentTerm, heartbeatTimeCall, grantedVotes, votedFor, state, shift, electionTimeCall
	global maybeNeedToCommit, matchIndex, nextIndex, log, recoveryMode, server_id, commitIndex
	if (state == 'l' and term >= currentTerm):
		if term > currentTerm:
			currentTerm = term
			state = 'f'
			grantedVotes = 0
			votedFor= None
			heartbeatTimeCall = False
			electionTimeCall = False
			shift = True
		elif success:
			matchIndex[followerId] += entriesToAppend
			maybeNeedToCommit = True
			if nextIndex[followerId] < log._length:
				msg = {'rpc':'appendEntries'
				, 'term':currentTerm
				, 'leaderId':server_id
				, 'prevLogIndex':nextIndex[followerId]-1
				, 'prevLogTerm':log._entries[nextIndex[followerId]-1]._term
				, 'entries': [x.__dict__ for x in log.slice(nextIndex[followerId], min(log._length, nextIndex[followerId]+100))]
				, 'leaderCommit':commitIndex}
				nextIndex[followerId]+=min(log._length,nextIndex[followerId]+100)-nextIndex[followerId]
				if nextIndex[followerId] == log._length:
					if recoveryMode:
						print("Follower %d log should be now in sync. Exiting recovery mode."%(followerId))
						recoveryMode = False
				sendMessage(followerId, msg)
		else:
			if recoveryMode == False: print("Follower %d log is outdated. Entering recovery mode."%(followerId))
			recoveryMode = True
			nextIndex[followerId] = entriesToAppend
			matchIndex[followerId] = nextIndex[followerId]-1
			if log._entries[nextIndex[followerId]-1] != None:
				msg = {'rpc':'appendEntries'
				, 'term':currentTerm
				, 'leaderId':server_id
				, 'prevLogIndex':nextIndex[followerId]-1
				, 'prevLogTerm':log._entries[nextIndex[followerId]-1]._term
				, 'entries': [log._entries[nextIndex[followerId]].__dict__]
				, 'leaderCommit':commitIndex}
				nextIndex[followerId] += 1
				sendMessage(followerId, msg)
			else: #Last log is behind oldest entry still in the log
				pass


#commitIndex - Start commit from this index
def commitEntries():
	global maybeNeedToCommit, commitIndex, clusterMembers, server_id
	global newCommitIndex, currentTerm, log, commitTime
	if maybeNeedToCommit == True:
		processEn = False
		#print "matchIndex for 2 = ", 	matchIndex[2]
		if maybeNeedToCommit:
			newCommitIndex = commitIndex-1
		while True:
			newCommitIndex += 1
			numReplicas = 1
			for node in clusterMembers:
				if node != server_id:
					if matchIndex[node] >= newCommitIndex: numReplicas+=1
			if numReplicas <= len(clusterMembers)/2:
				break
		print "newCommitIndex = ", newCommitIndex
		if log._entries[newCommitIndex-1]._term == currentTerm: #Possible faulty
			commitIndex = newCommitIndex
			processEn = True
		maybeNeedToCommit = False
		if processEn:
			processEntries(commitIndex) #Why commitIndex + 1?? #Possible fault
	threading.Timer(commitTime/1000.0, commitEntries).start()


def addEntry(requestId, requestData, clientId):
	global currentTerm, state, server_id, log, heartbeatTimeCall, lastKnownLeaderID
	global heartbeatTimeCall, shiftHeart, shift, electionTimeCall
	if state == 'l':
		entry = LogEntry(clientId, requestId, requestData, currentTerm)
		msg = {'rpc':'appendEntries'
		, 'term':currentTerm
		, 'leaderId':server_id
		, 'prevLogIndex': (log._length)-1
		, 'prevLogTerm' : log._entries[-1]._term
		, 'entries': [entry.__dict__]
		, 'leaderCommit':commitIndex};
		for node in clusterMembers:
			if node != server_id:
				if nextIndex[node] == log._length:
					nextIndex[node] += 1
					sendMessage(node, msg)
		log.push(entry.__dict__)
		heartbeatTimeCall = False
		shiftHeart = True
		electionTimeCall = False
		shift = True
	elif lastKnownLeaderID != None: #forward to leader
		msg = {'clientId':clientId
		, 'requestId' : requestId
		, 'requestData' : requestData
		, 'rpc':'addEntry'};
		sendMessage(lastKnownLeaderID, msg)

def readEntry(requestId, clientId, fileName):
	data = ""
	if os.path.exists(fileName) ==False:
		data = "Does not exist!"
	else:
		with open(fileName, "rb") as myfile:
			data = myfile.read()
	msg = {'rpc' : 'readEntryReply'
	, 'data' : data
	, 'requestId' : requestId};
	sendMessage(clientId, msg)

# def main():
temp = readFromPersistentStore()
if temp != None:
	currentTerm, votedFor, log, lastApplied = temp
# print temp
electionTimeout()
electionTimeCall = True
heartbeatTimeout()
persistentStoreTimeout()
commitEntries()

#Life
while True:
	message = receiver_socket.recv()
	sender_id, message = message.split(" ", 1)
	message = ast.literal_eval(message)
	print sender_id, message, type(message)
	rpc = message['rpc']
	if rpc == 'appendEntries':
		appendEntries(message['term'],
			message['leaderId'],
			message['prevLogIndex'],
			message['prevLogTerm'],
			message['entries'],
			message['leaderCommit'])
	elif rpc == 'requestVote':
		requestVote(message['term']
			, message['candidateId']
			, message['lastLogIndex']
			, message['lastLogTerm'])
	elif rpc == 'replyVote':
		replyVote(message['term'], message['voteGranted'])
	elif rpc == 'replyAppendEntries':
		replyAppendEntries(message['term']
			, message['followerId']
			, message['entriesToAppend']
			, message['success'])
	elif rpc == 'addEntry':
		addEntry(message['requestId']
			, message['requestData']
			, message['clientId'])
	elif rpc == 'readEntry':
		readEntry(message['requestId']
			, message['clientId']
			, message['fileName'])

# if __name__=="__main__":
# 	main()


utility.py
import pickle
import os

def writeToPersistentStore(currentTerm, votedFor, log, lastApplied):
	dic = {"currentTerm": currentTerm
	, "votedFor": votedFor
	, "log": log
	, 'lastApplied':lastApplied }
	with open("checkpoint.pkl", "wb") as outFile:
		pickle.dump(dic, outFile, pickle.HIGHEST_PROTOCOL)
		# pickle.dump(log, outFile, pickle.HIGHEST_PROTOCOL)

def readFromPersistentStore():
	newDict = {}
	if os.path.exists("checkpoint.pkl") == False:
		return None
	with open("checkpoint.pkl", "rb") as inFile:
		newDict = pickle.load(inFile)
		currentTerm = newDict['currentTerm']
		votedFor = newDict['votedFor']
		log = newDict['log']
		lastApplied = newDict['lastApplied']
	return currentTerm, votedFor, log, lastApplied

def write_to_file(msg):
	if msg is not None:
		clientId = msg['clientId']
		if os.path.isdir(clientId) == False:
			os.makedirs(clientId)
		path = clientId+"/"+msg['sourceId']
		with open(path, "a") as myfile:
			myfile.write("%s"%(msg['fileData']))


testPersistentstore.py
import pickle
class LogEntry(object):
	def __init__(self, clientId, requestId, data, term):
		self._clientId = clientId
		self._requestId = requestId
		self._data = data
		self._term = term

	def __repr__(self):
		return "\nclientId: "+ str(self._clientId) + \
		"\nrequestId: " + str(self._requestId) + \
		"\ndata: " + str(self._data) + \
		"\nterm: " + str(self._term) + "\n "


class Log(object):
	def __init__(self, server_id):
		self._firstIndex = 0
		self._length = 1;
		self._entries = [LogEntry(-1, 0, None, 0), ]

	def __repr__(self):
		return "\nfirstIndex " + str(self._firstIndex) + \
				"\nlength " + str(self._length) + \
				"\nEntries" + str(self._entries)

	def push(self, value):
		self._length += 1
		print value, type(value)
		trans = LogEntry(value['_clientId'], value['_requestId'], value['_data'], value['_term'])
		self._entries.append(trans)

	def pop(self):
		self._length -= 1
		self._entries.pop()

	def shift(self):
		pass

	def slice(self, from1, to):
		return self._entries[from1:to]

# currentTerm = 23
# votedFor = 1
# log = Log(1)
def writeToPersistentStore():
	global currentTerm, votedFor, log
	dic = {"currentTerm": currentTerm, "votedFor": votedFor, "log": log }
	with open("checkpoint.pkl", "wb") as outFile:
		pickle.dump(dic, outFile, pickle.HIGHEST_PROTOCOL)
		# pickle.dump(log, outFile, pickle.HIGHEST_PROTOCOL)

def readFromPersistentStore():
	global currentTerm, votedFor, log
	with open("checkpoint.pkl", "rb") as inFile:
		newDic = pickle.load(inFile)
		currentTerm = newDic['currentTerm']
		votedFor = newDic['votedFor']
		log = newDic['log']
		# log = pickle.load(inFile)

def main():
	# writeToPersistentStore()
	readFromPersistentStore()
	print "\ncurrentTerm: " + str(currentTerm)
	print "\nvotedFor: "+ str(votedFor)
	print "\nlog: " + str(log)

if __name__=="__main__":
	main()


clientServer.js
//node clientServer.js 5 tcp://127.0.0.1 12345 true 3000 5002
//0       1            2          3        4     5    6    7

var app = require('express')();
var bodyParser = require('body-parser');
var formidable = require('express-formidable');
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended:true}));
app.use(formidable());
var http = require('http').Server(app);
var zmq = require('zmq');


serverID = process.argv[2]
router_address = process.argv[3]
send_port_no = process.argv[4]
debug = (process.argv[5] == "true")
web_port = process.argv[6]
listen_portno = process.argv[7]


listen_socket = zmq.socket('pull');
listen_socket.connect(router_address+":"+listen_portno)

sender_socket = zmq.socket('push')
sender_socket.connect(router_address+":"+send_port_no)

var writePending_dict = [],
readPending_dict = [];


listen_socket.on('message', function(){

  // Parsing Message

  var args = Array.apply(null, arguments);
  console.log(args[0].toString('utf8'));
  var message = JSON.parse(args[0].toString('utf8'))
  console.log(message);
  if (message['dest'] != serverID)
    return;

  if (message['rpc'] == 'addEntryReply')
  	writePending_dict[message['requestId']] = message['status'];
  else if (message['rpc'] == 'readEntryReply'){
  	readPending_dict[message['requestId']] = message['data'];
  }

});


//For reading request
//Request Infromation:
//Post Request - multipart/formdata
//Keys - clientId, sourceId
app.post('/readData', function(req, res){
  var requestId = Math.random().toString(36).substring(7); //Assigning a randomID
  if (debug) console.log("Read RequestID assigned : " + requestId)
  readPending_dict[requestId] = 'failure';
  var data = req.fields;
  if (debug) console.log(data);
  var fileName = data['clientId']+"/"+data['sourceId'];
  msg = {'clientId' : serverID
  , 'requestId': requestId
  , 'rpc' : 'readEntry'
  , 'fileName' : fileName
  , 'customerId' : data['clientId']
  , 'sourceId' : data['sourceId']
  };
  sender_socket.send(JSON.stringify(msg));
  if (debug) console.log(msg);
  var count = 20;
  var my_int = setInterval(function(){
    if (count == 0){
      res.send("Server Timeout!");
      if (debug) "Request id " + requestId + "didn't got serve\n"
      clearInterval(my_int);
      return;
    }
    // count -= 1;
    if (readPending_dict[requestId] != "failure"){
      if (debug) console.log(readPending_dict[requestId]);
      res.send(readPending_dict[requestId]);
      clearInterval(my_int);
    }
  }, 100);
  return;
});


//For writing request
//Requests Information:-
//POST REQUEST - multipart/formdata
//Keys - fileData, clientId, sourceId
app.post('/pushData', function(req, res){
	var requestId = Math.random().toString(36).substring(7); //Assigning a randomID
	if (debug) console.log("Write RequestID assigned : " + requestId)
	writePending_dict[requestId] = 'failure';
  data = req.fields;
  if (debug) console.log(data)
  msg = {'clientId':serverID
  , 'requestId' : requestId
  , 'requestData' : {'fileData': new Date().getTime() + ":" + data['fileData'] + "\n", 'clientId': data['clientId'], 'sourceId':data['sourceId'] }
  , 'rpc':'addEntry'
  , 'customerId' : data['clientId']
  , 'sourceId' : data['sourceId']
  };
  sender_socket.send(JSON.stringify(msg));
  if (debug) console.log(msg)
  var count = 20;
  var my_int =   setInterval(function(){
    if (count == 0) {
    	res.send("Server Timeout!");
    	clearInterval(my_int);
  	  return;
    }
    // count -= 1;
    if (writePending_dict[requestId] == "Success"){
  	  if (debug) console.log("Log accepted");
  	  res.send("Log added!");
  	  clearInterval(my_int);
    }
  },100);
  return;
});


//Start http server
http.listen(web_port, function(){
  console.log('listening on *:3000');
});


